lang-ref/part1.md

```diff
@@ -0,0 +1,941 @@
+
+There is also a concept of semantic subtyping, which means a subset of the values of the basic types. For example, the byte type is a subset of integer as it allows only a subset of values that are a part of the integer value set.
+
+Additionally, Ballerina allows you to define your own types, such as arrays, maps and records, which are mutable.<br><br>
+
+## Unions
+
+Ballerina’s type system allows you some flexibility to define custom types based on the combination of two or more types. A union is, therefore, a superset of that combination.
+ 
+You have already seen this with the `‘T?’` notation where a variable can be of type T or nil.   
+
+Similarly, you can define a type that holds an integer as well as a string.
+
+```
+type flexType string|int;
+
+flexType a = 1;
+
+flexType a = “Hello”;
+```
+
+Using the `‘|’` symbol in the type definition, you can create a Union of multiple types. Therefore, in the above example, **``flexType``** can hold an integer value as well as a string value and the Ballerina compiler won’t complain.
+
+You can also apply the concept of union to structured data types like records. So it is possible to have a union of a record containing string type fields and a basic string type.
+
+```
+type StructuredName record {
+    string firstName;
+    string lastName;
+};
+
+type Name StructuredName|string;
+```
+ 
+At runtime, you can define the business logic by checking the currently held value of the union using the `is` operator.
+
+```
+function nameToString(Name nm) returns string {
+   
+    if nm is string {
+        return nm;
+        } else {
+        return nm.firstName + " " + nm.lastName;
+    }
+}
+```
+ 
+In this way, the ``is`` operator in a condition causes the declared type to be narrowed.<br><br>
+
+## Error Reporting
+
+Ballerina does not support the notion of exception. Instead, errors are handled as part of the normal control flow. Errors are first-class citizens in Ballerina. There are a few mechanisms for handling errors, which are centered around an error type. error is a basic type in Ballerina, and error values can be passed around just like any other type.
+
+Errors are reported by functions returning error values.
+
+```
+function parse(string s) returns int|error {
+    
+    int n = 0;
+    int[] cps = s.toCodePointInts();
+    
+    foreach int cp in cps {
+        int p = cp - 0x30;
+        if p < 0 || p > 9 {
+            return error("not a digit");
+        }
+        n = n * 10 + p;
+    }
+    return n;
+}
+```
+ 
+In the above code example, the function **``parse()``** returns a union of integer or error type. At the point of returning the error, the function *``error(“not a digit”)``*  is called.
+
+In case a function is expected to return only error values, a return type of ``error?`` is used.   Ignoring an error also has to be done explicitly.
```

Comment:
```suggestion
In case a function is expected to return only error values, a return type of ``error?`` is used. Ignoring an error also has to be done explicitly.
```